design-proposals/2023-06-17-ffi.md

```diff
@@ -0,0 +1,68 @@
+# Foreign Function Interface (FFI) Library
+
+## Motivation
+
+Enable FFI bindings (C API) with MapLibre API for Rust and Dart (Flutter) using a Foreign Function Interface (FFI) Library.
+
+## Proposed Change
+
+Add a C API to the MapLibre source tree in the form of a FFI library.  This library would be integrated into CMake as an optional target.  The external FFI API would be exposed using a C header file.  This header file would be consumed by Rust and Dart interface generation tools.
+
+A longer term strategy could be to expose extern C interfaces directly in the common source tree.  This would decrease the long term cost of ownership.
+
+The approach that evolved to this proposed solution involved decoupling GLFW from the GLFW app, then abstracting out platform and backend code.  This was initialy done targeting Flutter, after which the same API was used to create a Rust app that behaves similar to the GLFW application.
+
+### Common Interface
+
+* version - returns FFI interface version
+* initialize - returns unique context pointer
+* de_initialize - tears down map instance and frees context
+* draw_frame - renders a single frame
+* resize - resizes the map
+* run_task - used for housekeeping tasks independent of render path
+* load_gl_functions - function pointer callback for resolving GL functions
```

Comment:
This would need to be generalized since we have a Metal backend as well and soon a Vulkan backend.

Reply:
Yep.  This was just for GL at the time